packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -568,18 +663,73 @@ function updateHostText(current, workInProgress) {
   return null;
 }
 
+function resolveDefaultProps(Component, baseProps) {
+  if (Component && Component.defaultProps) {
+    // Resolve default props. Taken from ReactElement
+    const props = Object.assign({}, baseProps);
+    const defaultProps = Component.defaultProps;
+    for (let propName in defaultProps) {
+      if (props[propName] === undefined) {
+        props[propName] = defaultProps[propName];
+      }
+    }
+    return props;
+  }
+  return null;
+}
+
 function mountIndeterminateComponent(
   current,
   workInProgress,
+  Component,
   renderExpirationTime,
 ) {
   invariant(
     current === null,
     'An indeterminate component should never have mounted. This error is ' +
       'likely caused by a bug in React. Please file an issue.',
   );
-  const fn = workInProgress.type;
+
   const props = workInProgress.pendingProps;
+  if (
+    Component !== null &&
+    Component !== undefined &&
+    typeof Component.then === 'function'
```

Comment:
Maybe we should check `typeof Component === 'object'`? Seems like a function with a `.then` is ambiguous at best.